### PR TITLE
Unicode pubkeys

### PR DESCRIPTION
**Pull request by itkovian**

### CODE REVIEW

File: lib/vsc/administration/ldapsync.py [hunk 1]
```diff
@@ -121,7 +121,16 @@ def sync_altered_accounts(self, last, dry_run=True):
...
             logging.debug('fetching public key')
 
-            public_keys = [str(x.pubkey) for x in self.client.get_public_keys(account.vsc_id)]
+            public_keys = []
+            for something in self.client.get_public_keys(account.vsc_id):
+                logging.debug("Got pubkey %s for account %s", something.pubkey, account.vsc_id)
+                try:
+                    pk = str(something.pubkey)
+                except UnicodeEncodeError:
+                    pk = something.pubkey.encode('ascii', 'ignore')
```
Comment (wpoely86) [inline lib/vsc/administration/ldapsync.py:130]: add a test with a unicode error key?

Comment (itkovian) [reply lib/vsc/administration/ldapsync.py:130]: Well, not sure how to test since it is the actual adding in the LDAP that fails. But we need to mock that anyway. So I am not sure what adding some unicode test would add besides verifying that we get ascii.

File: lib/vsc/administration/ldapsync.py [hunk 2]
```diff
@@ -121,7 +121,16 @@ def sync_altered_accounts(self, last, dry_run=True):
                 logging.warning("Converting unicode to ascii for gecos resulting in %s", gecos)
             logging.debug('fetching public key')
 
-            public_keys = [str(x.pubkey) for x in self.client.get_public_keys(account.vsc_id)]
+            public_keys = []
+            for something in self.client.get_public_keys(account.vsc_id):
```
Comment (stdweird) [inline lib/vsc/administration/ldapsync.py:125]: `something`?

File: lib/vsc/administration/ldapsync.py [hunk 3]
```diff
@@ -121,7 +121,16 @@ def sync_altered_accounts(self, last, dry_run=True):
...
+                logging.debug("Got pubkey %s for account %s", something.pubkey, account.vsc_id)
+                try:
+                    pk = str(something.pubkey)
+                except UnicodeEncodeError:
+                    pk = something.pubkey.encode('ascii', 'ignore')
+                    logging.warning("Converting unicode to ascii for pubkey resulting in %s", pk)
```
Comment (stdweird) [inline lib/vsc/administration/ldapsync.py:131]: is there any value in trying to log the original key? because we are modifying the key
@itkovian but i'm curious what triggered this. isn't the `.pubkey` the actual key, so what unicode chars could be in it or are even allowed?

Comment (itkovian) [reply lib/vsc/administration/ldapsync.py:131]: There were unicode chars in the thingie after the key :)

**Comment by itkovian** [comment]
I have no idea how to create a unief test output for both py3 and py2. ensure_ascii_string does something differently in both cases.

**Comment by wpoely86** [comment]
We also need the call to `ensure_ascii_string` when writing out the ssh key in the user sync:

```
2020-06-08 15:43:11,578 ERROR      sync_vsc_users.vsc.administration.user MainThread  Cannot process user vsc10277
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/vsc/administration/user.py", line 514, in process_users
    user.populate_home_dir()
  File "/usr/lib/python2.7/site-packages/vsc/administration/user.py", line 400, in populate_home_dir
    [p.pubkey for p in self.pubkeys])
  File "/usr/lib/python2.7/site-packages/vsc/filesystem/posix.py", line 398, in populate_home_dir
    fp.write("\n".join(ssh_public_keys + ['']))
UnicodeEncodeError: 'ascii' codec can't encode character u'\u201c' in position 725: ordinal not in range(128)

```

**Comment by wpoely86** [comment]
@itkovian Can't you make py2-only and py3-only test?

**Comment by itkovian** [comment]
@wpoely86 Dunno, do you have an example on how to do that?

**Comment by wpoely86** [comment]
```
from vsc.utils.py2vs3 import is_py2, is_py3

@skipIf(is_py2(), "Python 3 only")
```

Doesn't that work?

**Comment by wpoely86** [comment]
I can't merge it but it looks good to me.

**Comment by wdpypere** [comment]
well, now stuff does not get tested under python3. so I would expect a separate test for py3?

File: test/ldapsync.py [hunk 4]
```diff
@@ -104,6 +104,26 @@ def test_sync_altered_accounts(self, mock_add_or_update):
         ldap_attrs = {'status': ['active'], 'scratchDirectory': ['/user/scratch/gent/vsc400/vsc40075'], 'dataDirectory': ['/user/data/gent/vsc400/vsc40075'], 'cn': 'vsc40075', 'homeQuota': ['5242880'], 'institute': ['gent'], 'loginShell': ['/bin/bash'], 'uidNumber': ['2540075'], 'researchField': ['Bollocks'], 'gidNumber': ['2540075'], 'gecos': ['Foo Bar'], 'dataQuota': ['1'], 'homeDirectory': ['/user/home/gent/vsc400/vsc40075'], 'mail': ['foobar@ugent.be'], 'scratchQuota': ['1'], 'pubkey': ['pubkey1', 'pubkey2'], 'instituteLogin': ['foobar'], 'uid': ['vsc40075']}
         mock_add_or_update.assert_called_with(VscLdapUser, test_account.vsc_id, ldap_attrs, True)
 
+
+    @skipIf(is_py3(), "Python 2 only")
```
Comment (stdweird) [inline test/ldapsync.py:108]: why is there no py3 test?

Comment (itkovian) [reply test/ldapsync.py:108]: Because we may assume that ensure_ascii_string is tested in vsc-install and that this will not affect the code flow or its operation?

Comment (itkovian) [reply test/ldapsync.py:108]: Adding a test for py3 will only test that this specific function also works in py3, which is not something we need to do here?

Comment (stdweird) [reply test/ldapsync.py:108]: then what are you testing in py2? or what is different in py2 vs py3?
i'm confused why this exception is needed at all

Comment (itkovian) [reply test/ldapsync.py:108]: I am testing that the unicode stuff is replaced as expected.

Comment (wdpypere) [reply test/ldapsync.py:108]: so that should be tested on python3 and python2 no? I know the order differs (I had that on other repos as well) . But either make a test that works on both, or make 2 tests.

**Comment by itkovian** [comment]
Adding a test for py3 makes py2 fail.

File: test/ldapsync.py [hunk 5]
```diff
@@ -104,6 +104,46 @@ def test_sync_altered_accounts(self, mock_add_or_update):
...
+        mock_client.account[test_account.vsc_id].quota.get.return_value = (200, test_quota)
+
+        mock_add_or_update.return_value = UPDATED
+        ldapsyncer = LdapSyncer(mock_client)
+        accounts = ldapsyncer.sync_altered_accounts(1)
+        self.assertEqual(accounts, {'error': set([]), 'new': set([]), 'updated': set([test_account.vsc_id])})
+        ldap_attrs = {'status': ['active'], 'scratchDirectory': ['/user/scratch/gent/vsc400/vsc40075'], 'dataDirectory': ['/user/data/gent/vsc400/vsc40075'], 'cn': 'vsc40075', 'homeQuota': ['5242880'], 'institute': ['gent'], 'loginShell': ['/bin/bash'], 'uidNumber': ['2540075'], 'researchField': ['Bollocks'], 'gidNumber': ['2540075'], 'gecos': ['Foo Bar'], 'dataQuota': ['1'], 'homeDirectory': ['/user/home/gent/vsc400/vsc40075'], 'mail': ['foobar@ugent.be'], 'scratchQuota': ['1'], 'pubkey': ['some pubkey \\u201chuppelde@daar.com\\u201d'], 'instituteLogin': ['foobar'], 'uid': ['vsc40075']}
+        mock_add_or_update.assert_called_with(VscLdapUser, test_account.vsc_id, ldap_attrs, True)
+
+    @skipIf(is_py2(), "Python 3 only")
```
Comment (stdweird) [inline test/ldapsync.py:127]: i have a hard time diffing this copy/paste code? maybe put a `if/else block` around the parts that are actually different?

File: test/ldapsync.py [hunk 6]
```diff
@@ -104,6 +104,27 @@ def test_sync_altered_accounts(self, mock_add_or_update):
...
+        mock_client.account.modified[1].get.return_value = (200, [test_account_1])
+        mock_client.account[test_account.vsc_id].usergroup.get.return_value = (200, test_usergroup_1)
+        mock_client.get_public_keys.return_value = [mkVscAccountPubkey(p) for p in test_unicode_pubkeys]
+        mock_client.account[test_account.vsc_id].quota.get.return_value = (200, test_quota)
+
+        mock_add_or_update.return_value = UPDATED
+        ldapsyncer = LdapSyncer(mock_client)
+        accounts = ldapsyncer.sync_altered_accounts(1)
+        self.assertEqual(accounts, {'error': set([]), 'new': set([]), 'updated': set([test_account.vsc_id])})
+        if is_py2():
```
Comment (stdweird) [inline test/ldapsync.py:122]: how about the actual different key(s)?

File: test/sync_vsc_email.py [hunk 7]
```diff
@@ -81,9 +81,8 @@ def test_sync(self, pl):
         newmap = [
             "vsc40001@vscentrum.be huppel@bar.com",
             "vsc40002@vscentrum.be foo2@bar.com",
-            "",
         ]
 
         vps.do(False)
         with open(cmfn, 'r') as c:
-            self.assertEqual(c.read(), "\n".join(newmap))
+            self.assertEqual(c.readlines(), newmap)
```
Comment (stdweird) [inline test/sync_vsc_email.py:88]: readlines does not strip the newline char (i think)

Comment (itkovian) [reply test/sync_vsc_email.py]: indeed. Fixed it :)

**Comment by stdweird** [comment]
@itkovian woohoo, perseveranc++ ;)
